proposals/0436-recordwildcards-total.rst

```diff
@@ -0,0 +1,102 @@
+----------
+
+Some functions require all the fields of a record to be used, or at least considered, e.g. ``hash`` or ``toJSON``, or even something more security-facing. Currently, if one uses ``RecordWildCards`` and they add a field to the record later on, nothing will tell the developer to take a look at these functions.
+
+::
+
+ data Foo = Foo { a :: Int, b :: Bool }
+ 
+ instance ToJSON Foo where
+   toJSON Foo{..} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..} = hash (a, b)
+ 
+ -- adding a 'c' field does not raise any indication of needing to look at `toJSON` or `hash`
+
+Existing workarounds are described in `this blog post <https://cs-syd.eu/posts/2021-09-10-undefined-trick>`_, all of which leave something to be desired:
+
+* Positional arguments break when fields are reordered, which matters in functions like `toJSON`
+* Using `undefined` is a bit of a hack, plus it fails `-Wunused-pattern-binds`. It also fails to explicitly say which field is unused (which isn't strictly necessary, but it's nice to have)
+
+The `safe-wild-cards <https://hackage.haskell.org/package/safe-wild-cards>`_ package provides a Template Haskell solution to this, but it would be nice to support this natively.
+
+One could even think of this as the dual of the `Or pattern proposal <https://github.com/ghc-proposals/ghc-proposals/pull/43>`_, to warn on unused terms in product types as opposed to unused terms in sum types.
+
+Proposed Change Specification
+-----------------------------
+
+When ``-XRecordWildCards`` is enabled, in addition to supporting ``Foo{..}`` in a pattern match, also support ``Foo{..!}``, which desugars to ``Foo{a = a, b = b}``.
+
+Examples
+--------
+
+Now, the example in the Motivation section could be written as
+
+::
+
+ instance ToJSON Foo where
+   toJSON Foo{..!} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..!} = hash (a, b)
+   
+Notice the lack of any change, except for swapping ``Foo{..}`` for ``Foo{..!}``.
+
+
+Effect and Interactions
+-----------------------
+The desugared syntax would raise an ``-Wunused-matches`` warning if a new field is added but not used.
+
+As far as I know, the proposed ``Foo{..!}`` syntax doesn't conflict with any current syntax.
+
+
+Costs and Drawbacks
+-------------------
+Since it's just syntax sugar, development "shouldn't" take long. Likewise, I don't see this adding much maintenance cost.
+
+It's completely optional, so novice users can avoid this syntax completely.
+
+Primary drawback is updating tooling/editor support for the new syntax.
+
+
+Alternatives
+------------
+
+* Add an additional extension, e.g. ``-XRecordWildCardsStrict``, that implies ``-XRecordWildCards`` and enables this extension. I'm not opposed to this, but since the change doesn't break existing ``-XRecordWildCards`` syntax, I don't see a reason to break out yet another extension.
+
```

Comment:
Expected to see this proposal compared to [linear record selectors](https://gitlab.haskell.org/ghc/ghc/-/issues/18570). Those would seem to provide the same benefit of making it impossible to forget using a field.

Reply:
I wasn't aware that was an open proposal. For next time, how would I know what proposals to address when making a proposal?

I'm also not sure what linear record selectors has to do with this. It seems to only apply to records with only one field? IIUC it would also only work if the function was linear on the record data type, but this proposal is more general than linear types and should work for non-linear functions as well. e.g.
```hs
data Foo = Foo { a :: Int, b :: Int }

foo :: Foo -> Int
foo Foo{..!} = a + a + b
```
should work also (not sure why you'd want to do this, but my point is linearity is orthogonal to this proposal)